Purchases/Purchasing/StoreKit1/ProductsFetcherSK1.swift

```diff
@@ -113,32 +142,43 @@ extension ProductsFetcherSK1: SKProductsRequestDelegate {
     }
 
     func request(_ request: SKRequest, didFailWithError error: Error) {
+        defer {
+            self.cancelRequestToPreventTimeoutWarnings(request)
+        }
+
         queue.async { [self] in
             Logger.appleError(Strings.storeKit.skproductsrequest_failed(error: error))
-            guard let products = self.productsByRequests[request] else {
+
+            guard let productRequest = self.productsByRequests[request] else {
                 Logger.error(Strings.purchase.requested_products_not_found(request: request))
                 return
             }
-            guard let completionBlocks = self.completionHandlers[products] else {
-                Logger.error(Strings.purchase.callback_not_found_for_request(request: request))
-                self.productsByRequests.removeValue(forKey: request)
-                return
-            }
 
-            self.completionHandlers.removeValue(forKey: products)
-            self.productsByRequests.removeValue(forKey: request)
-            for completion in completionBlocks {
-                completion(Set())
+            if productRequest.retriesLeft <= 0 {
+                guard let completionBlocks = self.completionHandlers[productRequest.identifiers] else {
+                    Logger.error(Strings.purchase.callback_not_found_for_request(request: request))
+                    self.productsByRequests.removeValue(forKey: request)
+                    return
+                }
+
+                self.completionHandlers.removeValue(forKey: productRequest.identifiers)
```

Comment:
That slightly changes the behavior from what it was before (this code is unchanged in this PR): it would be removing the request from `productsByRequest` before making sure a `completionHandler` exists... Probably not an issue, but I'm afraid of making such a subtle change together in this much larger change.